### PR TITLE
[WIP] Auto joins

### PR DESCRIPTION
**Pull request by jpreese**
This is the supporting UI for auto joins. Marked as WIP as we'll still need to implement new message tokens on the server to support adding/removing joins.

UI: http://imgur.com/a/W9SGG

### CODE REVIEW

**Comment by jpreese** [comment]
Food for thought, we discussed this ad nauseum earlier today.

The TL;DR of the chat log is essentially do we need to introduce an entirely new interface to manage autojoins, or can we just leverage the honorary member/squad system.

1) A squad has to exist anyway in order to manage a channel (setting info, permissions, etc)
2) If you're going to be autojoining a channel, you're going to be in that channel.. a lot. You practically are an "Honorary Member".
3) It works great for channels like Europe and hock, where you want to screen those who are allowed in the channel anyway. Hock only wants registered users, Europe prefers Euros.

When you bring everything together, its more.. do you want autojoins to be seen as "Honorary Members" or do you want them to be their own thing.

I am personally leaning a little towards the honorary member system, instead of introducing new UI / database schema / socket end points. But obviously needed your input/final say.

Chat log from this mornings discussion:

[14:16]Q8ball:
you still only get one 'main' squad

[14:16]gran:
its like 20 people already
[14:16]Q8ball:
which is the tag that'll show

[14:16]Inter:
thats neat

[14:16]Q8ball:
its probably how the old system worked tbh, we just didn't know

[14:16]gran:
and you can set it not to show your squad tag
[14:17]Inter:
cool, thanks

[14:27]Frog:
hmm

[14:29]Frog:
that gets me thinking, I wonder if we could just leverage the Honorary Member functionality for autojoins. We were going to implement them as their own thing (http://imgur.com/a/W9SGG), but I wonder if its actually needed. trying to think of cases where you wouldnt want them together

[14:29]Frog:
my immediate thought was hock league.. but honestly, you have to register for that anyway. so joining the "squad" as an honorary member, could be your registration

[14:34]Q8ball:
hmm ya

[14:36]Frog:
there has to be a hock squad anyway to administrate the channel, set the info, etc

[14:37]Frog:
the only case I can think of would be if you just wanted a place to hangout.. it seems a little strange to have to create a squad just for that, but I anticipate it being such a small use case

[14:37]Frog:
and we can avoid adding in new screens/support for something that doesn't have a large use

[14:37]Q8ball:
that's why europe channel exists already

[14:37]Frog:
(open to opinions)

[14:37]Q8ball:
and VET channel

[14:38]Q8ball:
thats actually why i asked you in steam if you could create a squad for me and try to set it as owner

[14:38]Q8ball:
b/c i couldn't create a new room that i could manage

[14:38]gran:
i assumed thats what honorary member was there for anyway, since its only ability is to join locked channels, same as an invite would
[14:38]Frog:
yeah and I feel like those are good examples of honorary members, you dont want people really joining unless you want them to be there

[14:38]Q8ball:
yup

[14:38]gran:
the only thing i can think of is that you cannot create 2 squads
[14:38]Frog:
speaking of which, theres a vet channel?!

[14:38]gran:
so atm if i wanted to create another locked channel i couldnt
[14:39]Q8ball:
i was going to create one if there wasn't going to be one implemented

[14:39]Q8ball:
so any old uniballer (anyone prior to steam release) could hang out

[14:40]Q8ball:
similar to stan's europe channel

[14:40]Frog:
right, but at least for now, channel administration is done through squads, and I dont see that changing. so a squad has to exist anyway

[14:40]Q8ball:
for all euro friends

[14:40]Frog:
so it is a limitation to not be able to make 2 squads, but I dont know if its a deal breaker

[14:41]gran:
its not
[14:42]gran:
still i think allowing 2 per user would be ok
[14:43]gran:
but its a thing that affects v few people
[14:44]Frog:
yeah, I see the use case though.. say you, gran, wanted to make the Europe channel for all of the euro guys, but you wanted to own your own squad, its not currently possible

[14:44]gran:
ah and the one difference is that you have to be invited as a honorary member
[14:44]Frog:
so you'd have to find someone else to make Europe, so you could have your own

[14:44]gran:
whereas autojoins you can manage yourself
[14:44]Frog:
yeah

[14:44]Q8ball:
yup

[14:45]Frog:
which I dont know if thats entirely a bad thing.. I mean.. if youre going to be autojoining

[14:45]Frog:
that means youll be there, a lot

[14:45]Frog:
so.. youre an honorary member ;x

[14:45]gran:
yea but you need a person with rights to be online
[14:45]gran:
and ask them
[14:45]gran:
and for them to see it and accept
[14:46]Frog:
yeah, I think itd just depend on the channel. I mean, I think thats perfectly ok for say.. Europe. If you only want Euros, then there you go, you have to be given exclusive rights to have it as a join

[14:46]Frog:
same with VET

[14:47]Frog:
but it might be a slight inconvenience for just open to the public hangout type channels or something

[14:49]Frog:
can you don the squad tag as an honorary member?

[14:49]gran:
no
[14:49]Frog:
ok good

[14:49]Frog:
yeah, dunno

[14:51]gran:
it's just a little counter intuitive for new players to have to go to squad/join/apply to squad in order to autojoin a channel
[14:51]Frog:
I just dont see us implementing a different channel management system other than squads, so while it seems slightly strange to create a squad JUST for a hangout channel, thats probably how its going to be.

[14:51]Frog:
yeah

[14:51]gran:
i mean you saw above that i had to explain to inter that its not a "real" squad
[14:52]zoron:
Maybe a simple fix would be, when a /help file is created it brings up a second window that explains what you are trying to get across?
[14:54]Frog:
the TL;DR of all of this is.. is it worth adding an interface for autojoins, or can we just get away with/does it make sense for honoary member to be your autojoin, since the squad has to exist anyway, for the channel to exist

[14:54]zoron:
right
[14:54]Q8ball:
right

[14:55]Frog:
part of me leans on honorary member fits a ton of cases anyway, the biggest issue is the whole VET/Europe/etc channels AND wanting to have your own.. but thats not really an autojoin issue

[14:55]Frog:
thats just steam forces us into one account now

[14:55]zoron:
i think it would be easier to make autojoins part of being an honorary member
[14:55]Frog:
so we cant make a new account to make a new squad for us, we have to juggle it somewhow

[14:56]zoron:
you can be owner of multiple squads?
[14:56]zoron:
or have ownership rights
[14:56]Frog:
I think gran was saying you cant

[14:56]Q8ball:
ya, just make a button for request autojoin, and invite auto join that leverages the honorary member as their permissions 

[14:56]Frog:
so if Q8 makes VET, he also couldnt own RKO

[14:57]Q8ball:
via context menu 

[14:57]Frog:
of if Sergo made Europe, he also couldnt own Kiu, etc

[14:57]zoron:
correct. i just tried it
[14:58]Frog:
so a solution could be to be able to own multiple squads.. but it does seem a -little- strange to call "hangout" channels, squads.

[14:58]Frog:
*shrug*

[14:58]zoron:
so, im assuming that there is a statement that is preventing you from creating multiple squads
[14:58]Frog:
but we dont have another system for it, when all is said and done, they are squads

[14:59]zoron:
the other thing is, would this create a database problem
[14:59]zoron:
say you allow people to own more than one squad
[14:59]zoron:
you have a person come in and own  50 squads.. thats 50 lines..
[14:59]zoron:
at some point it is going to create an error
[14:59]gran:
2 is reasonable
[14:59]zoron:
each squad needs its own permissions
[15:00]Frog:
mmm... shouldnt, there really isnt a different between owning a squad and being a member of it as far as I'm aware. It's just a permission value 1 - 5

[15:00]Frog:
ah I see, shrug. maybe. doubt it.

[15:01]gran:
keeping autojoins in the db as honorary members as opposed to local autojoins will take more db space than the few people who decide to have 2 squads
[15:01]Frog:
as gran says though, capping it is within reason

[15:01]zoron:
right, just needs a cap
[15:01]Frog:
well, they wouldnt be local since you'd want them to work on the web client too

[15:01]gran:
ah ok

**Comment by jpreese** [comment]
Follow up discussion with Nick, makes me lean fully towards the non squad approach. Players will be able to leverage bots to manage channels without a backing squad.
